Make local_directory a required argument for spilling impls

wence- opened the pull request:
For automated cleanup when the cluster exits, the on-disk spilling directory needs to live inside the relevant worker's local_directory. Since we do not have a handle on the worker when constructing the keyword arguments to DeviceHostFile or ProxifyHostFile, instead take advantage of dask/distributed#7153 and request that we are called with the worker_local_directory as an argument. Closes #1018.

wence- commented:
This is a breaking change (since it is no longer possible to provide a `None` local_directory to the various spilling classes and get some kind of default). On the other hand, as #1018 demonstrates, there is a need to put stuff in the appropriate worker's local directory (and that location can still be controlled by configuration via usual dask mechanisms).

wence- commented:
cc @madsbk as implementor of most of these classes.

pentschev reviewed (commented):
Left a few questions, but otherwise looks good. Also would be good to wait for @madsbk 's review as well, in case he notices anything that may not play well with proxied/JIT-Unspill.

madsbk reviewed (commented):
LGTM. 
I know this has not been changed in this PR, but I would suggest that we use the `tmp_path` fixture instead of `tempfile.TemporaryDirectory()`.

wence- commented:
> LGTM, but do we need a fallback in the case Distributed doesn't provide the `worker_local_directory` argument?

Ah, if the user is running with an older version of distributed. Yes, good point.

> Also, I know this has not been changed in this PR, but I would suggest that we use the `tmp_path` fixture instead of `tempfile.TemporaryDirectory()`.

I will have a look.

pentschev commented:
> > LGTM, but do we need a fallback in the case Distributed doesn't provide the `worker_local_directory` argument?
> 
> Ah, if the user is running with an older version of distributed. Yes, good point.

I don't think we need one. We are always moving up the Distributed requirement in Dask-CUDA, i.e., we don't provide backwards compatibility.

wence- commented on dask_cuda/cuda_worker.py:
```diff
@@ -175,7 +175,6 @@ def del_pid_file():
                         device_memory_limit, device_index=i
                     ),
                     "memory_limit": memory_limit,
-                    "local_directory": local_directory,
```
No, I think this is one we want the user to be able to control. The point with the plugins is that each worker, when it boots makes something relative to `local_directory / dask-worker-space / worker-SOMERANDOMSTRING` and we don't know what `SOMERANDOMSTRING` is (but we do still want control over `local_directory`.

wence- commented on dask_cuda/local_cuda_cluster.py:
```diff
@@ -289,7 +289,6 @@ def __init__(
                     {
                         "device_memory_limit": self.device_memory_limit,
                         "memory_limit": self.memory_limit,
-                        "local_directory": local_directory,
```
As above, I think this is OK.

wence- commented:
> I will have a look.

Done. Via a module-scope fixture.

pentschev commented:
It seems tests are failing because [Distributed nightly gets installed](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L969), which is [posteriorly overwritten by the current release](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L2,272). This happens because [Distributed nightly is installed in ](https://github.com/rapidsai/dask-cuda/blob/3e5a19b246aa684d2bedfb4673960d895d81dd24/ci/gpu/build.sh#L72-L76)build.sh, and I guess that [Dask-CUDA’s conda package that is later built and installed](https://github.com/rapidsai/dask-cuda/blob/branch-22.12/ci/gpu/build.sh#L93-L104) grabs Distributed stable again. I'm asking internally why this is happening this way.

wence- commented:
> It seems tests are failing because [Distributed nightly gets installed](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L969), which is [posteriorly overwritten by the current release](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=driver-450,LINUX_VER=centos7,PYTHON=3.8/968/consoleFull#L2,272). This happens because [Distributed nightly is installed in ](https://github.com/rapidsai/dask-cuda/blob/3e5a19b246aa684d2bedfb4673960d895d81dd24/ci/gpu/build.sh#L72-L76)build.sh, and I guess that [Dask-CUDA’s conda package that is later built and installed](https://github.com/rapidsai/dask-cuda/blob/branch-22.12/ci/gpu/build.sh#L93-L104) grabs Distributed stable again. I'm asking internally why this is happening this way.

Thanks @pentschev!

pentschev commented:
rerun tests

pentschev commented:
Failing tests are unrelated, opened an issue in https://github.com/rapidsai/dask-cuda/issues/1032 .

pentschev commented:
rerun tests

wence- commented:
These look like actually my fault, let me fix.

wence- commented:
> These look like actually my fault, let me fix.

Actually perhaps not, can't reproduce locally.

pentschev commented:
@wence- just gave it a quick try and I can reproduce the failure using `docker.io/gpuci/rapidsai:22.12-cuda11.2-devel-centos7-py3.8 ` (CI test that fails) but not `docker.io/gpuci/rapidsai:22.12-cuda11.5-devel-ubuntu20.04-py3.9` (one of the CI tests that pass). The difference seems to hint at something with Python 3.8, but not sure to be honest.

wence- commented:
> @wence- just gave it a quick try and I can reproduce the failure using `docker.io/gpuci/rapidsai:22.12-cuda11.2-devel-centos7-py3.8 ` (CI test that fails) but not `docker.io/gpuci/rapidsai:22.12-cuda11.5-devel-ubuntu20.04-py3.9` (one of the CI tests that pass). The difference seems to hint at something with Python 3.8, but not sure to be honest.

Thanks, I think I fixed this. It was due to use of `typing.MutableMapping` which breaks `inspect.getfullargspec` in Python 3.8. I've implemented some workarounds.

pentschev reviewed (approved):
This is great, thanks @wence- for figuring out the issue and for the fix. LGTM.

pentschev commented:
@gpucibot merge

wence- commented on dask_cuda/device_host_file.py:
```diff
@@ -234,6 +235,34 @@ def __init__(
         # For Worker compatibility only, where `fast` is host memory buffer
         self.fast = self.host_buffer if memory_limit is None else self.host_buffer.fast
 
+    if sys.version_info > (3, 8):
```
Because 3.8.1 > 3.8, but fixed in #1036
